ships/0022-build-strategy-volumes.md

```diff
@@ -0,0 +1,265 @@
+
+- Securing volume usage/access with admission control.
+- Allow developers to inject arbitrary volumes into `BuildRun`s.
+
+## Proposal
+
+### User Stories
+
+#### Share information in build steps
+
+As a build strategy author
+I want to declare volumes in my build steps
+So that I can share information across build steps
+
+#### Use persistent volumes in build runs
+
+As a developer
+I want to declare persistent volume claims in my Builds
+So that I can cache resources like image layers and artifacts for subsequent BuildRuns
+
+### Implementation Notes
+
+The `BuildStrategy` API will be enhanced so that volumes can be declared alongside the build strategy steps.
+The volumes will declare a default source for the underlying filesystem.
+Each build step must opt into using the volume as a normal container volume mount.
+Volumes can be marked `optional` indicating that the volume is not needed for a `BuildRun` to execute.
+Volumes can also indicate if the source can be overrode by a `Build` or `BuildRun`.
+
+Both the `Build` and `BuildRun` APIs will be enhanced to allow volumes to be specified.
+The volume name referenced in the `Build` or `BuildRun` must be declared in the build strategy.
+If a `BuildRun` references a volume that does not exist (either directly or in its parent `Build` object), the build should fail.
+
+#### Deprecate Implicit emptyDir Volumes
+
+Shipwright currently creates an implicit `emtpyDir` volume if one or more build steps declare a volume mount.
+This behavior should be deprecated as a prerequsite to releasing this feature.
+Implicit emptyDir volumes can then be removed when this feature is released.
+
+#### Strategy Volumes API
+
+`BuildStrategy`:
+
+```yaml
+metadata:
+  name: cached-docker-build
+spec:
+  buildSteps:
+  - name: build
+    image: quay.io/my-org/my-builder:latest
+    volumeMounts:
+    - name: build-metadata
+      mountPath: /home/build/metadata
+    - name: image-cache
+      mountPath: /var/lib/containers/cache
+  volumes:
+  - name: build-metadata
+    description: "Build metadata"
+    optional: true
+    volumeSource:
+      type: EmptyDir # Type discriminator, this wil let us support new volume sources over time.
+      emptyDir: {}
+  - name: image-cache
+    description: "Container image cache"
+    volumeSource:
+      overridable: true # indicates the volume source can be different in a Build or BuildRun
+      type: EmptyDir
+      emptyDir: {}
+```
+
+`Build` and `BuildRun`:
+
+```yaml
+spec:
+  ...
+  strategy:
+    name: cached-docker-build
+    kind: BuildStrategy
+  volumes:
+  - name: image-cache
+    volumeSource:
+      type: PersistentVolumeClaim # When overriding, the type can be changed
+      persistentVolumeClaim:
+        name: pvc-image-cache
+```
+
+#### TaskRun Generation
+
+When Shipwright generates the `TaskRun` for the `BuildRun`, volumes are created as follows:
+
+0. For each volume in the build strategy, a corresponding `volumeMount` is added to each build step.
```

Comment:
This does not match what you have in your example. There, the volumeMount is defined in the build strategy. And, I am in favor of asking the build strategy author to do declare it explicitly, because

- the volumeMount is the place where the mountPath is defined
- I see use cases when it comes to secret volumes where one would not necessarily want it to be mounted at all places